KV watcher panics out when NATS connection unstable

Seeing a scenario while performing Unit tests for side inputs where the KV watcher go routine panics out with null pointer exception. In the test the Nats connection is closed towards the end of the test but the watcher routine is still running and we're seeing the issue. Need to debug further to understand the exact issue.

Steps to reproduce the behavior:
The test is flaky and does not reproduce everytime, can try to run the test multiple times. 
Seeing in 

1. `pkg/sideinputs/initializer/initializer_test.go`
2. `pkg/sideinputs/synchronizer/synchronizer_test.go`

**Expected behavior**
We should gracefully error out instead of panic

**Screenshots**
```
/usr/local/go/bin/go tool test2json -t /Users/skohli/Library/Caches/JetBrains/GoLand2023.1/tmp/GoLand/___TestSideInputsInitializer_Success_in_github_com_numaproj_numaflow_pkg_sideinputs_initializer.test -test.v -test.paniconexit0 -test.run ^\QTestSideInputsInitializer_Success\E$
=== RUN   TestSideInputsInitializer_Success
{"level":"info","ts":"2023-08-09T11:02:12.203322-07:00","logger":"numaflow","caller":"initializer/initializer.go:109","msg":"Initializing Side Input data for %q/tmp/side-input/TEST"}
{"level":"info","ts":"2023-08-09T11:02:12.203331-07:00","logger":"numaflow","caller":"jetstream/kv_watch.go:140","msg":"watcher initialization and subscription got nil value","pipeline":"testPipeline","kvBucketName":"sideInputTestWatch"}
{"level":"info","ts":"2023-08-09T11:02:12.204522-07:00","logger":"numaflow","caller":"initializer/initializer.go:109","msg":"Initializing Side Input data for %q/tmp/side-input/TEST2"}
{"level":"warn","ts":"2023-08-09T11:02:13.216557-07:00","logger":"numaflow","caller":"jetstream/kv_watch.go:134","msg":"Failed to stop the watcher","pipeline":"testPipeline","kvBucketName":"sideInputTestWatch","watcher":"sideInputTestWatch","error":"nats: connection closed"}
{"level":"info","ts":"2023-08-09T11:02:13.216679-07:00","logger":"numaflow","caller":"jetstream/kv_watch.go:136","msg":"Succeeded to recreate the watcher, since the channel is closed","pipeline":"testPipeline","kvBucketName":"sideInputTestWatch"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x10169a8bc]

goroutine 31 [running]:
github.com/numaproj/numaflow/pkg/shared/kvs/jetstream.(*jetStreamWatch).Watch.func1()
	/Users/skohli/numa/numaflow/pkg/shared/kvs/jetstream/kv_watch.go:117 +0x7c
created by github.com/numaproj/numaflow/pkg/shared/kvs/jetstream.(*jetStreamWatch).Watch
	/Users/skohli/numa/numaflow/pkg/shared/kvs/jetstream/kv_watch.go:102 +0x15c


Process finished with the exit code 1
```

**Environment (please complete the following information):**

- Kubernetes: [e.g. v1.18.6]
- Numaflow: [e.g. v0.5.1]


---

--- a/pkg/shared/kvs/jetstream/kv_watch.go
+++ b/pkg/shared/kvs/jetstream/kv_watch.go
@@ -96,11 +96,14 @@ func (k kvEntry) Operation() kvs.KVWatchOp {
 // Watch watches the key-value store (aka bucket).
 func (jsw *jetStreamWatch) Watch(ctx context.Context) (<-chan kvs.KVEntry, <-chan struct{}) {
 	var err error
+	// create a new watcher, it will keep retrying until the context is done
+	// returns nil if the context is done
 	kvWatcher := jsw.newWatcher(ctx)
 	var updates = make(chan kvs.KVEntry)
 	var stopped = make(chan struct{})
 	go func() {
-		for {
+		// if kvWatcher is nil, it means the context is done
+		for kvWatcher != nil {
 			select {
 			case <-ctx.Done():
 				jsw.log.Infow("stopping WatchAll", zap.String("watcher", jsw.GetKVName()))

